### PR TITLE
Trigger print via state change

### PR DESCRIPTION
**Pull request by steeevio**
I was struggling to put the `<ReactToPrint>` within sibling / child components to the `<ComponentToPrint>` component. 

Maybe there is a better way of doing this but for me it seemed to make it more flexible if I could trigger print via state. then I can do it with a button from anywhere.

I just added a some of props `<ReactToPrint>` that allow the user to decide if they want to trigger from state
`useState={true}`

check the current print state - set to true to run print
`statePrint= {true}`

user should add a function that sets the print state back to false after use.
`handleStatePrintOff={this.handleStatePrintOff} `

### CODE REVIEW

File: example/index.js [hunk 1]
```diff
@@ -5,6 +5,13 @@ import ReactToPrint from '../src';
 import ComponentToPrint from './ComponentToPrint';
 
 class Example extends React.Component {
+  
+  constructor(props){
+    super(props);
+    this.state = {
+      print:false,
```
Comment (MatthewHerbst) [inline example/index.js]: Nit:
```suggestion
      print: false,
```

File: example/index.js [hunk 2]
```diff
@@ -25,14 +32,27 @@ class Example extends React.Component {
     this.componentRef = ref;
   }
 
+  handleStatePrint = (e)=>{
```
Comment (MatthewHerbst) [inline example/index.js]: Nit:

```suggestion
  handleStatePrint = (e) => {
```

File: example/index.js [hunk 3]
```diff
@@ -25,14 +32,27 @@ class Example extends React.Component {
...
+    this.setState({print : true});
+  }
+  handleStatePrintOff = (e)=>{
```
Comment (MatthewHerbst) [inline example/index.js]: Nit:
```suggestion
  handleStatePrintOff = (e) => {
```

File: example/index.js [hunk 4]
```diff
@@ -25,14 +32,27 @@ class Example extends React.Component {
...
+    this.setState({print : false});
```
Comment (MatthewHerbst) [inline example/index.js]: Nit:
```suggestion
    this.setState({ print: false });
```

File: example/index.js [hunk 5]
```diff
@@ -25,14 +32,27 @@ class Example extends React.Component {
     this.componentRef = ref;
   }
 
+  handleStatePrint = (e)=>{
+    this.setState({print : true});
```
Comment (MatthewHerbst) [inline example/index.js]: Nit:
```suggestion
    this.setState({ print: true });
```

File: example/index.js [hunk 6]
```diff
@@ -25,14 +32,27 @@ class Example extends React.Component {
...
       <div>
+        <button onClick={this.handleStatePrint} type="button">Print with state!</button>
         <ReactToPrint
           trigger={this.renderTrigger}
           content={this.renderContent}
           onBeforePrint={this.handleBeforePrint}
           onAfterPrint={this.handleAfterPrint}
+          statePrint={this.state.print}
+          handleStatePrintOff={this.handleStatePrintOff}
+          useState={false}
```
Comment (MatthewHerbst) [inline example/index.js]: The default value of `useState` should be `false`, so you should never need to pass the prop with `false`

File: src/index.js [hunk 7]
```diff
@@ -53,6 +53,7 @@ class ReactToPrint extends React.Component {
   }
 
   handlePrint = () => {
+
```
Comment (MatthewHerbst) [inline src/index.js]: Nit:
```suggestion
```

File: src/index.js [hunk 8]
```diff
@@ -177,6 +178,9 @@ class ReactToPrint extends React.Component {
       }
     };
 
+
```
Comment (MatthewHerbst) [inline src/index.js]: Nit: please remove these extra newlines

File: src/index.js [hunk 9]
```diff
@@ -189,10 +193,31 @@ class ReactToPrint extends React.Component {
...
-      ref: this.setRef,
-    });
+    if(this.props.useState === true){
+      
+      if(this.props.statePrint === true){
+        this.handlePrint()
+      } 
+
+      return (
+        <div></div>
```
Comment (MatthewHerbst) [inline src/index.js]: Please use `<div />`. Also, please make sure to add a semicolon after the closing `)`

File: src/index.js [hunk 10]
```diff
@@ -189,10 +193,31 @@ class ReactToPrint extends React.Component {
       trigger,
     } = this.props;
 
-    return React.cloneElement(trigger(), {
-      onClick: this.handlePrint,
-      ref: this.setRef,
-    });
+    if(this.props.useState === true){
+
```
Comment (MatthewHerbst) [inline src/index.js]:
```suggestion
```

File: src/index.js [hunk 11]
```diff
@@ -189,10 +193,31 @@ class ReactToPrint extends React.Component {
...
+    if(this.props.useState === true){
+      
+      if(this.props.statePrint === true){
+        this.handlePrint()
+      } 
+
+      return (
+        <div></div>
+      )     
+    }else{
```
Comment (MatthewHerbst) [inline src/index.js]:
```suggestion
    } else {
```

File: src/index.js (hunk 9 above)
Comment (MatthewHerbst) [reply src/index.js]: Also, why did you choose to return an empty `div` here rather than `null`?

File: src/index.js [hunk 12]
```diff
@@ -189,10 +193,31 @@ class ReactToPrint extends React.Component {
...
+      
+      return React.cloneElement(trigger(), {
+        onClick: this.handlePrint, 
+        ref: this.setRef,
+      });
+    }
+
```
Comment (MatthewHerbst) [inline src/index.js]: Nit: please remove extra newlines

File: src/index.js [hunk 13]
```diff
@@ -189,10 +193,31 @@ class ReactToPrint extends React.Component {
       trigger,
     } = this.props;
 
-    return React.cloneElement(trigger(), {
-      onClick: this.handlePrint,
-      ref: this.setRef,
-    });
+    if(this.props.useState === true){
```
Comment (MatthewHerbst) [inline src/index.js]: Please destructure this prop above

File: src/index.js [hunk 14]
```diff
@@ -189,10 +193,31 @@ class ReactToPrint extends React.Component {
...
+      
+      if(this.props.statePrint === true){
```
Comment (MatthewHerbst) [inline src/index.js]: Please destructure this prop above

**Review by MatthewHerbst — CHANGES_REQUESTED** [review]
Thanks for the PR! A few quick things first:

1. I left a bunch of style related comments on the PR that would be great if you could clean up
2. There shouldn't have been any changes to `package-lock.json` with this PR. You likely did an `npm install` instead of `npm ci`. Please do not commit an updated `package-lock.json` with this PR
3. Please add the new props to the `propTypes` definitions

I think overall what you are really trying to do is solve #68, which wants a way to print without having to use the button. While I think this approach is a good bandaid attempt to that before the work for #68 is complete, the issue is that if/when we do the work for #68, removing these changes would be a breaking change which should ideally be avoided. 

Would you be interested in working on #68 together rather than attempting this approach?

File: example/index.js (hunk 1 above)
Comment (steeevio) [reply example/index.js]: Hi, this was my first stab at a pull request, so thanks for taking the time over the feedback! I've a had a go at changing it as required - hopefully I got everything, if not let me know.

I'm new to react so not really sure how much help I can be in dealing with #68, I'm happy to try but honestly you might just end up spending more time explaining things to me.

Comment (MatthewHerbst) [reply example/index.js]: Pretty great stab for your first PR, well done!

I actually think #68 doesn't have much to do with React - the goal is to take code that's currently baked into the component and move it into its own standalone JS. Then the React component can depend on it, and we can make it available to others. For example, a sample export might be:

```js
import { print } from 'react-to-print'
```

Then you could just call `.print(<someReactRef>|<someDOMNode>)`

That would solve what you are trying to do here, which is control when you print via external code. I'll take another look at it to see what a good path forward should be.

Do note I'm not the package owner, just an active contributor.

**Comment by gregnb** [comment]
hey @MatthewHerbst really appreciate all your help. I've added push access for you

**Comment by MatthewHerbst** [comment]
@gregnb thank you! Looking forward to getting some free time to address these issues :)

File: example/index.js (hunk 1 above)
Comment (steeevio) [reply example/index.js]: Thanks! I see what you mean, yes ultimately that seems like a better solution.

**Comment by MatthewHerbst** [comment]
Going to close this in favor of #88 

@steeevio would love any feedback you have about that PR, which is still in early stages though the meat of it is done

**Pull request closed without merging**